radix-engine/src/model/substates/substate.rs

```diff
@@ -266,3 +261,337 @@ impl Into<GlobalAddressSubstate> for Substate {
+            }
+            SubstateRef::ComponentState(substate) => {
+                let scrypto_value = ScryptoValue::from_slice(&substate.raw).unwrap();
+                (scrypto_value.global_references(), scrypto_value.node_ids())
+            }
+            SubstateRef::KeyValueStoreEntry(substate) => {
+                let maybe_scrypto_value = substate
+                    .0
+                    .as_ref()
+                    .map(|raw| ScryptoValue::from_slice(raw).unwrap());
+                if let Some(scrypto_value) = maybe_scrypto_value {
+                    (scrypto_value.global_references(), scrypto_value.node_ids())
+                } else {
+                    (HashSet::new(), HashSet::new())
+                }
+            }
+            SubstateRef::NonFungible(substate) => {
+                let maybe_scrypto_value = substate
+                    .0
+                    .as_ref()
+                    .map(|non_fungible| ScryptoValue::from_typed(non_fungible));
+                if let Some(scrypto_value) = maybe_scrypto_value {
+                    (scrypto_value.global_references(), scrypto_value.node_ids())
+                } else {
+                    (HashSet::new(), HashSet::new())
+                }
+            }
+            _ => (HashSet::new(), HashSet::new()),
+        }
+    }
+}
+
+pub fn verify_stored_value_update(
+    old: &HashSet<RENodeId>,
+    missing: &HashSet<RENodeId>,
+) -> Result<(), RuntimeError> {
+    // TODO: optimize intersection search
+    for old_id in old.iter() {
+        if !missing.contains(&old_id) {
+            return Err(RuntimeError::KernelError(KernelError::StoredNodeRemoved(
+                old_id.clone(),
+            )));
+        }
+    }
+
+    for missing_id in missing.iter() {
+        if !old.contains(missing_id) {
+            return Err(RuntimeError::KernelError(KernelError::RENodeNotFound(
+                *missing_id,
+            )));
+        }
+    }
+
+    Ok(())
+}
+
+pub struct SubstateRefMut<'f, 's, R: FeeReserve> {
+    lock_handle: LockHandle,
+    prev_children: HashSet<RENodeId>,
+    node_pointer: RENodePointer,
+    offset: SubstateOffset,
+    call_frames: &'f mut Vec<CallFrame>,
+    track: &'f mut Track<'s, R>,
+}
+
+impl<'f, 's, R: FeeReserve> SubstateRefMut<'f, 's, R> {
+    pub fn new(
+        lock_handle: LockHandle,
+        node_pointer: RENodePointer,
+        offset: SubstateOffset,
+        prev_children: HashSet<RENodeId>,
+        call_frames: &'f mut Vec<CallFrame>,
+        track: &'f mut Track<'s, R>,
+    ) -> Result<Self, RuntimeError> {
+        let substate_ref_mut = Self {
+            lock_handle,
+            prev_children,
+            node_pointer,
+            offset,
+            call_frames,
+            track,
+        };
+        Ok(substate_ref_mut)
+    }
+
+    pub fn offset(&self) -> &SubstateOffset {
+        &self.offset
+    }
+
+    pub fn flush(&mut self) -> Result<(), RuntimeError> {
```

Comment:
Should we consume the `RefMut` upon `flush`? It seems flush can only be done once.

Reply:
hrm, good idea! flush could be called multiple times currently but probably simpler to only allow it once.